binary-install/tests/all/utils/mod.rs

```diff
@@ -0,0 +1,73 @@
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use std::fs::{File, OpenOptions};
+use std::io::{self, Read, Write};
+use std::net::TcpListener;
+use std::thread;
+use std::time::Duration;
+
+pub const TEST_SERVER_HOST: &'static str = "localhost";
+
+pub fn start_server(port: u32, tarball: Option<Vec<u8>>) -> thread::JoinHandle<TcpListener> {
+    let handle = thread::spawn(move || {
+        let listener = TcpListener::bind(format!("{}:{}", TEST_SERVER_HOST, port)).unwrap();
+
+        for stream in listener.incoming() {
+            let mut stream = stream.unwrap();
+
+            let mut buffer = [0; 512];
+
+            stream.read(&mut buffer).unwrap();
+
+            let response = "HTTP/1.1 200 OK\r\n\r\n";
+
+            stream.write(response.as_bytes()).unwrap();
+
+            match tarball.to_owned() {
+                Some(tar) => {
+                    stream.write(tar.as_ref()).unwrap();
+                }
+                None => {}
+            }
+
+            stream.flush().unwrap();
+        }
+        listener
+    });
+
+    // Wait for server to start...
+    thread::sleep(Duration::from_secs(1));
```

Comment:
Perhaps use a [channel](https://doc.rust-lang.org/nightly/std/sync/mpsc/fn.channel.html) where the server can send a message after it has bound the tcp listener? That would be less racy.

Reply:
Sounds like a plan, will try this out. Thanks!